### ISSUE TITLE
HttpDownloader fails to download on IPv6 network

### ISSUE THREAD

**Issue by damienmg** [bazelbuild/bazel#2486]
TL;DR: We failed to demo tensorflow build at fosdem on the recent infra because IPv6 only.

**Comment by abergmeier-dsfishlabs** [comment]
<sub>that is embarassing</sub> SCNR

**Comment by philwo** [comment]
I'll try to reproduce and fix on Monday. Currently don't have an IPv6 network available for testing.

**Comment by philwo** [comment]
I've built a little HttpURLConnection playground and tried downloading a file from a dual-stack host (bazel-mirror.storage.googleapis.com) with either IPv4/IPv6 enabled or IPv6-only.

I verified that I can download the file with "curl" in all network setups.

Even though this is supposed to "just work", it horribly fails with Java for all IPv6 related scenarios (ie. dual-stack or IPv6-only):

Trying to download from IPv4 host with dual-stack connectivity: Works.
Trying to download from IPv6 host with dual-stack connectivity: "No route to host" exception.
Trying to download from IPv4 host with IPv6-only: "Network unreachable" (but HttpURLConnection doesn't retry with IPv6)
Trying to download from IPv4 host with dual-stack connectivity: "No route to host" exception.

I'm not sure what's going on here and will investigate more.

**Comment by philwo** [comment]
This seems relevant: http://stackoverflow.com/questions/29103828/cant-connect-to-ipv6-only-host-from-java

tl;dr - a bug in the JRE lets it pick a wrong interface (awdl0) for IPv6 sockets on macOS. The workaround is to disable AirDrop via "sudo ifconfig awdl0 down". A fix would probably involve writing our own SocketFactory that skips over the bad interface and using that SocketFactory for all connections.

**Comment by philwo** [comment]
Unassigning, because I'm not working in this area.

**Comment by danieldanciu** [comment]
I think somebody should take a look at this. It's a bit embarrassing that Bazel basically doesn't work with ipv6 addresses, and for us, that means either enable ipv4 in our network or use another build system.

**Comment by philwo** [comment]
@danieldanciu Can you give me some more details about your environment? Have you seen this on macOS only, or also on other platforms? Are you running an IPv6-only network with DNS64 and NAT64?

@buchgr Any idea how we could fix / workaround this? Last time I investigated this, it looked like an OpenJDK bug only on macOS, related to OpenJDK picking the wrong interface (awdl0) for IPv6 communication.

**Comment by danieldanciu** [comment]
Sure. The platform is indeed a Mac, and the problem is that java.net.Socket is basically not working with ipv6 addresses on a Mac, as described here:
 http://stackoverflow.com/questions/29103828/cant-connect-to-ipv6-only-host-from-java

There is no workaround on the new macs (with a touch bar).

The fix on the Bazel side would be to use java.nio.SocketChannel instead of java.net.Socket whenever making an http connection. I tried to fix this myself, but I couldn't find where Bazel is using the Socket, I suspect it may be using it indirectly through the grpc problem.

The only way to work around this was to actually edit the libnet.dylib library with a hex editor and alter the code such that the correct interface is picked up.

**Comment by philwo** [comment]
Thank you for the detailed description and pointers! I’ll make sure that this gets fixed.

**Comment by buchgr** [comment]
The HttpDownloader seems to use HttpURLConnection which most likely uses java.net.Socket under the hood. I don't think that one can just pass in his own socket.

> I suspect it may be using it indirectly through the grpc problem.

what are you refering to?

**Comment by danieldanciu** [comment]
Typo, I meant "grpc library". But you're likely right that the socket is opened via HttpURLConnection. Unfortunately, the socket factory only allows creating java.net.Sockets, so you guys may need to write your own HttpUrlConnection that uses socketchannel. Although I am pretty sure someone at Google already wrote this. :)

**Comment by sergiocampama** [comment]
This just bit me as well, any chance it could be looked at as corporate infrastructure is moving into IPv6?

**Comment by danieldanciu** [comment]
What version of basel/java/mac are you using? The problem went away for me, and I think it was the upgrade to Bazel 0.28 that did it.

**Comment by or-shachar** [comment]
I just got the same issue on bazel 0.29.1

**Comment by DFrenkel** [comment]
We have a build machine that has IPv6 network capable of reaching out to external destinations and IPv4 address that is internal-only. We are also getting Network unreachable exceptions in our build, but I believe I have a workaround that is based on a [documented](https://docs.oracle.com/javase/8/docs/technotes/guides/net/ipv6_guide/index.html#ipv6-related) way in Java to prefer IPv6 by setting **java.net.preferIPv6Addresses** system property. 

Specifically you can do the following:

* Use `--host_jvm_args=-Djava.net.preferIPv6Addresses=true` [startup option](https://docs.bazel.build/versions/master/user-manual.html#startup_options), for example by having the following line in your `bazelrc`: `startup --host_jvm_args=-Djava.net.preferIPv6Addresses=true`
* If you are using a mixture of Java and non-java build targets, and your Java tests need to download dependencies as well, also use `--jvmopt` [tool flag](https://docs.bazel.build/versions/master/user-manual.html#flags-options), for example by having the following like in your `bazelrc`: `build --jvmopt=-Djava.net.preferIPv6Addresses`
* If you are using [rules_jvm_external](https://github.com/bazelbuild/rules_jvm_external) as well (for example, `coursier_fetch` for maven artifact version resolution), add `-Djava.net.preferIPv6Addresses=true` to an environment variable called `COURSIER_OPTS` to [Provide JVM options for Coursier](https://github.com/bazelbuild/rules_jvm_external#provide-jvm-options-for-coursier-with-coursier_opts)

**Comment by philwo** [comment]
I just tried to reproduce this and embarrassingly, this issue is still present - Bazel 3.4.1 on macOS 10.15.6 fails to download files via HTTP when the computer is in an IPv6-only (with NAT64 and DNS64) network. 😬 

```
$ bazel build //src:bazel
Starting local Bazel server and connecting to it...
INFO: Repository bazel_toolchains instantiated at:
  no stack (--record_rule_instantiation_callstack not enabled)
Repository rule http_archive defined at:
  /Users/philwo/src/bazel/tools/build_defs/repo/http.bzl:336:31: in <toplevel>
WARNING: Download from https://mirror.bazel.build/github.com/bazelbuild/bazel-toolchains/releases/download/3.1.0/bazel-toolchains-3.1.0.tar.gz failed: class java.net.ConnectException Network is unreachable (connect failed)
WARNING: Download from https://github.com/bazelbuild/bazel-toolchains/releases/download/3.1.0/bazel-toolchains-3.1.0.tar.gz failed: class java.net.ConnectException Network is unreachable (connect failed)
ERROR: An error occurred during the fetch of repository 'bazel_toolchains':
   java.io.IOException: Error downloading [https://mirror.bazel.build/github.com/bazelbuild/bazel-toolchains/releases/download/3.1.0/bazel-toolchains-3.1.0.tar.gz, https://github.com/bazelbuild/bazel-toolchains/releases/download/3.1.0/bazel-toolchains-3.1.0.tar.gz] to /private/var/tmp/_bazel_philwo/7a01905b4627ca044e5e3f5ad5b14d26/external/bazel_toolchains/bazel-toolchains-3.1.0.tar.gz: Network is unreachable (connect failed)
ERROR: no such package '@bazel_toolchains//rules': java.io.IOException: Error downloading [https://mirror.bazel.build/github.com/bazelbuild/bazel-toolchains/releases/download/3.1.0/bazel-toolchains-3.1.0.tar.gz, https://github.com/bazelbuild/bazel-toolchains/releases/download/3.1.0/bazel-toolchains-3.1.0.tar.gz] to /private/var/tmp/_bazel_philwo/7a01905b4627ca044e5e3f5ad5b14d26/external/bazel_toolchains/bazel-toolchains-3.1.0.tar.gz: Network is unreachable (connect failed)

$ wget https://mirror.bazel.build/github.com/bazelbuild/bazel-toolchains/releases/download/3.1.0/bazel-toolchains-3.1.0.tar.gz
--2020-07-23 15:10:58--  https://mirror.bazel.build/github.com/bazelbuild/bazel-toolchains/releases/download/3.1.0/bazel-toolchains-3.1.0.tar.gz
Resolving mirror.bazel.build (mirror.bazel.build)... 64:ff9b::82d3:16eb
Connecting to mirror.bazel.build (mirror.bazel.build)|64:ff9b::82d3:16eb|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 850795 (831K) [application/x-gzip]
Saving to: ‘bazel-toolchains-3.1.0.tar.gz’

bazel-toolchains-3.1.0.tar.gz    100%[=========================================================>] 830.85K  --.-KB/s    in 0.04s   

2020-07-23 15:10:58 (18.8 MB/s) - ‘bazel-toolchains-3.1.0.tar.gz’ saved [850795/850795]
```

"Luckily" our work environment apparently migrated the WiFi to be IPv6-only, so that we can now easily test and work on this issue soonish.

**Comment by sgowroji** [comment]
Hi there! We're doing a clean up of old issues and will be closing this one. Please reopen if you’d like to discuss anything further. We’ll respond as soon as we have the bandwidth/resources to do so.

**Comment by philwo** [comment]
I’m pretty sure that this isn’t fixed yet.

@meisterT Could you try whether Bazel works on an IPv6 only network these days?

**Comment by kaycebasques** [comment]
I think I just hit this snag while working on a gLinux in a Google office. After switching to our legacy IPv4 network the issue resolved.

**Comment by seanptmaher** [comment]
Yeah, I think I'm facing it right now not even on gLinux, but random arch linux on GoogleGuest. Fixed on GoogleGuest-IPV4.

thanks for making siso work off-corp @philwo  :)

**Comment by philwo** [comment]
This bug was mistakenly closed. It's not fixed yet. Can you reopen it? @meteorcloudy

**Comment by jvolkman** [comment]
I can reproduce this always sitting at my desk with my MacBook plugged into the (apparently IPv6-only) Google corporate wired network.

Java 9+ supports `-Djava.net.preferIPv6Addresses=system` ([bug](https://bugs.openjdk.org/browse/JDK-8016521), [commit](https://hg.openjdk.org/jdk9/jdk9/jdk/rev/3a4e4503049b)), which defers to the OS and ~~[RFC 3484](https://datatracker.ietf.org/doc/html/rfc3484)~~ [RFC 6724](https://datatracker.ietf.org/doc/html/rfc6724) for selecting an appropriate endpoint address rather than always prefering either IPv4 or IPv6. This seems like the correct approach.

I've solved the problem so far on my machine with the following changes

In `~/.bazelrc`
```
startup --host_jvm_args=-Djava.net.preferIPv6Addresses=system
```

In `~/.zshrc` (required for Coursier, etc. in rules_jvm_external)
```
# Make Java work nicely with IPv6
export JAVA_TOOL_OPTIONS="-Djava.net.preferIPv4Stack=false -Djava.net.preferIPv6Addresses=system"
```

Can we make Bazel default to `-Djava.net.preferIPv6Addresses=system`? `--host_jvm_args` is still around to override, if necessary. And a similar change could be made to `rules_jvm_external`.

**Comment by fmeum** [comment]
@jvolkman Nice find, this seems to be the right approach. Would you be willing to send a PR that adds these args somewhere around here? https://github.com/bazelbuild/bazel/blob/4c5a8027cc22cb02892abe0168c812b07c6e3447/src/main/cpp/blaze.cc#L447

**Comment by jvolkman** [comment]
I think that #29237 will improve the situation. I've tested the `system` value on the corp wired network (ipv6-only), corp wireless (dual stack, I think), and at home (ipv4 only) and, at least on my macbook, all three environments worked.

But these legacy Java http functions are still deficient. `HttpURLConnection` only attempts a connection to the first result returned from `getaddrinfo()`, and if it fails, tough luck. More modern approaches either iterate over the result list and use the first one that works, or use the more complicated [Happy Eyeballs](https://en.wikipedia.org/wiki/Happy_Eyeballs) approach which fires off parallel requests (after a short delay) and uses the first winner. There's a nice overview of the problem at https://bugs.openjdk.org/browse/JDK-8170568.

The most robust solution to this problem would be to swap `HttpURLConnection` with [okhttp](https://square.github.io/okhttp/) or similar, which is well tested (IIUC it is Android's default http library) and implements Happy Eyeballs. ~~An older version of `okhttp` is even included with Bazel - I think it's used for gRPC.~~ (it's listed in https://github.com/bazelbuild/bazel/blob/master/maven_install.json but I didn't find it in the actual installed files)

**Comment by fmeum** [comment]
Could we use Java's HttpClient instead? In JDK 26, which we can update to pretty much now, this even supports HTTP3. I set an agent on that migration in https://github.com/bazelbuild/bazel/pull/28675, but haven't found the time to fix the remaining test failures.

**Comment by ngocson2vn** [comment]
> startup --host_jvm_args=-Djava.net.preferIPv6Addresses=system

@jvolkman This works. Thank you!

**Comment by meisterT** [comment]
What's left to do here?

**Comment by jvolkman** [comment]
As described in https://github.com/bazelbuild/bazel/issues/2486#issuecomment-4200095997 I think it'd still be beneficial to switch Bazel's downloader from `HttpURLConnection` to something more modern (the built-in HttpClient, or something that implements RFC 8305). Maybe that can be a separate FR.
